teps/0018-default-bundle.md

```diff
@@ -0,0 +1,165 @@
+authors:
+  - "@coryrc"
+creation-date: 2020-09-15
+last-updated: 2020-09-15
+status: in pr review
+---
+
+# TEP-0018 Allow a Run to Specify a Default Bundle
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Observability](#observability)
+  - [New Variables](#new-variables)
+  - [Explicit Bundles from in-repo Resources](#explicit-bundles-from-in-repo-resources)
+- [Test Plan](#test-plan)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+In a Pipeline, a Task can be referenced by name. This name is currently looked
+up in the current namespace where the Pipeline and PipelineRun exist. It should
+be possible for a PipelineRun or TaskRun to specify a bundle to be searched for
+a match instead.
+
+## Motivation
+
+When CI is defined in-repo, some method must be provided to test Tekton objects
+during presubmit without altering Tasks being used by others. Additionally
+support multiple repositories and branches in a single namespace without the
+possibility of name conflict.
+
+### Goals
+
+Provide a different place to lookup Tekton objects when referenced solely by name.
+
+### Non-Goals
+
+- Override explicitly-labeled referenced Tasks (i.e. if they have taskRef.bundle).
+  This means further work is needed for presubmit testing of bundles generated
+  from resources defined in the repo-under-test. [more](#explicit-bundles-from-in-repo-resources)
+
+## Requirements
+
+## Proposal
+
+Create a new key for PipelineRun and TaskRun called `spec.lookup.bundle`; use it
+for finding referenced names in Tasks and Pipelines.
+
+### User Stories
+
+#### Story 1
+
+I have a repository where I have the entire CI/CD system using Tekton
+resources defined in the repository. I wish to test commits to the repository
+(presubmits) before allowing them to merge. I wish to support using the exact
+resources when running locally.
+
+My Pipelines refer to Tasks defined in-repo by name only and same for
+PipelineRun->Pipeline and TaskRun->Task. I can still use `_Ref.bundle` if I
+am referring to resources which are not created by this branch of the repository.
+
+My presubmit test takes non-Run resources and creates a bundle. All PipelineRuns
+and TaskRuns are modified to use the bundle as the default lookup environment. I
+can now execute in the production namespace without interfering with the
+environment in the existing namespace.
+
+#### Story 2
+
+I wish to support multiple repositories with CI/CD defined in-repo in the same
+namespace without the possibility of name conflicts. The system creating
+the PipelineRun must always specify a default bundle.
```

Comment:
The tool (described above) could be used in such scenario, a bit like what is currently done in our tekton as code experiment — where we create a new namespace, here we would run that tool to pack and push the bundle and execute the `*Run` with it.